diff --git a/node_modules/zen/src/mesh.js b/node_modules/zen/src/mesh.js
index b578834..74850bc 100644
--- a/node_modules/zen/src/mesh.js
+++ b/node_modules/zen/src/mesh.js
@@ -83,17 +83,19 @@ function Mesh(root) {
         console.STAT &&
           console.STAT(+new Date(), msg.length, "# on hear batch");
         var P = opt.puff;
+        var offset = 0;
         function go() {
           var S = +new Date();
           var i = 0,
             m;
-          while (i < P && (m = msg[i++])) {
+          while (i < P && (m = msg[offset + i])) {
             mesh.hear(m, peer);
+            i++;
           }
-          msg = msg.slice(i); // slicing after is faster than shifting during.
+          offset += i;
           console.STAT && console.STAT(S, +new Date() - S, "hear loop");
           flush(peer); // force send all synchronously batched acks.
-          if (!msg.length) {
+          if (offset >= msg.length) {
             return;
           }
           puff(go, 0);
